### ISSUE TITLE
XGBoostError: Check failed: gptr.size() != 0 && gptr.back() == info.labels.size() group structure not consistent with #rows

### ISSUE THREAD

**Issue by h3n0r1k** [dmlc/xgboost#3359]
Hello,

i know this issue has been posted already but cant find any solution. 

> XGBoostError: [15:46:54] d:\build\xgboost\xgboost-0.71.git\src\objective\rank_obj.cc:51: Check failed: gptr.size() != 0 && gptr.back() == info.labels.size() group structure not consistent with #rows

Using the following python code:
`import xgboost as xgb
import pandas as pd

##########

dtrain = xgb.DMatrix('svmlightTrainXG.txt')
dtest = xgb.DMatrix('svmlightTestXG.txt')

grouptrain = pd.read_table("../train.group.txt", header=None, sep='\n', engine='python')
grouptest = pd.read_table("../test.group.txt", header=None, sep='\n', engine='python')

dtrain.set_group(grouptrain)
dtest.set_group(grouptest)

##########

param = {'max_depth': 2, 'eta': 1, 'silent': 1, 'objective': 'rank:pairwise', 'label_offset':-1}
param['nthread'] = 4
param['eval_metric'] = 'auc'

num_round = 1000
bst = xgb.train(param, dtrain, num_round)

bst.save_model('0001.model')

ypred = bst.predict(dtest, ntree_limit=bst.best_ntree_limit)`

Error appears at line: 

> bst = xgb.train(param, dtrain, num_round)

Group file looks like this:

> 2545
1521
3217
15083
3966
1316
1584
3881
7991
14922
1181
43363
3250
1504
322
1953
12708
12705
8201
8192

And has a combined size of 149405L, wich is the same size i get when using dtrain.num_row() -> 149405L.

Feature file example:

> 0 0:16.57408832550151 1:13.66846065669936 2:16.57408832550151 3:5.203462337638929 4:6 5:2 6:6 7:1 8:0.09412942378936597 9:0.03093519066422871 10:0.08247045245342387 11:0.09894571165380825 12:9367 13:2.807840531572873 14:2.615091036984734 15:2.807840531572873 16:1.649324238179496 17:1.791759469228055 18:0.6931471805599453 19:1.791759469228055
0 0:2.856522914272229 1:12.72698948219145 2:2.856522914272229 4:2 5:2 6:2 8:0.03144164350186064 9:0.03897795800131735 10:0.0221058679706962 12:11782 13:1.049605120950391 14:2.543724894592617 15:1.049605120950391 17:0.6931471805599453 18:0.6931471805599453 19:0.6931471805599453
0 0:10.81109569901811 1:26.24536613051975 2:10.81109569901811 3:8.760923452679947 4:4 5:2 6:4 7:2 8:0.02489626282980397 9:0.07156681077701474 10:0.05583603923977173 11:0.1723383117778112 12:9738 13:2.380572986275721 14:3.267489445093515 15:2.380572986275721 16:2.170301316364918 17:1.386294361119891 18:0.6931471805599453 19:1.386294361119891 20:0.6931471805599453

Thanks for helping

**Comment by Shiki-H** [comment]
@h3n0r1k i ran into the same problem as you. If I understand you correctly, did you mean that your group file has the same number of lines as your dtrain? I can't find the source of the documentation for file format now, but if I remember correctly, the numbers in group file should be the number of documents in the same query, and you need to align your training data so that all instances within a same query are in the same chunk. For example, if your train looks like this (qid should not be in your dmatrix, I am just including it here for your reference)

| qid  | label  | features  |
|---|---|---|
| 1  | 0  | (features)  |
| 1  |  1 | (features)  |
| 1  | 0  | (features)  |
| 2  | 0  | (features)  |
| 2  | 1  | (features)  |
| 2  | 1  | (features)  |
| 2  | 1  | (features)  |

Then your group file should be 
3
4

In short, the sum of numbers in your group array should equal to the number of rows in your dtrain matrix. At least this helped me solve the problem.
